yt/frontends/gadget/data_structures.py

```diff
@@ -54,39 +52,142 @@ def _byte_swap_32(x):
+        # Format 2?
+        elif rhead == 8:
+            return 2, '<'
+        elif rhead == _byte_swap_32(8):
+            return 2, '>'
+        else:
+            raise RuntimeError(
+                "Gadget snapshot file is likely corrupted! "
+                "The first 4 bytes represent %s (as little endian int32). "
+                "But we are looking for %s (for format 1) or 8 (for format 2)."
+                % (rhead, first_header_size)
+            )
+
+    @property
+    def position_offset(self):
+        """Offset to the position block."""
+        n_header = len(self.size)
+        offset = 8 * n_header + sum(self.size)
+        if self.gadget_format[0] == 2:
+            offset += SNAP_FORMAT_2_OFFSET * (n_header + 1)
+        return offset
+
+    @property
+    def size(self):
+        """Header size in bytes."""
+        def single_header_size(single_header_spec):
+            return sum(field[1] * np.dtype(field[2]).itemsize
+                       for field in single_header_spec)
+        return [single_header_size(spec) for spec in self.spec]
+
+    @property
+    def value(self):
+        """Header values as a dictionary."""
+        # The entries in this header are capitalized and named to match Table 4
+        # in the GADGET-2 user guide.
+        gformat, endianswap = self.gadget_format
+        # Read header
+        with self.open() as f:
+            hvals = {}
+            for spec in self.spec:
+                if gformat == 2:
+                    f.seek(f.tell() + SNAP_FORMAT_2_OFFSET)
+                hvals_new = read_record(f, spec, endian=endianswap)
+                hvals.update(hvals_new)
+        # Remove placeholder keys
+        for key in self._placeholder_keys:
+            if key in hvals:
+                del hvals[key]
+        # Convert length 1 list to scalar value
+        for i in hvals:
+            if len(hvals[i]) == 1:
+                hvals[i] = hvals[i][0]
+        return hvals
 
+    def open(self):
+        """Open snapshot file."""
+        for filename in [self.filename, self.filename + '.0']:
+            if os.path.exists(filename):
+                return open(filename, 'rb')
+        raise RuntimeError("Snapshot file %s does not exist." % self.filename)
 
-def _get_gadget_format(filename, header_size):
-    # check and return gadget binary format with file endianness
-    ff = open(filename, 'rb')
-    (rhead,) = struct.unpack('<I', ff.read(4))
-    ff.close()
-    if (rhead == _byte_swap_32(8)):
-        return 2, '>'
-    elif (rhead == 8):
-        return 2, '<'
-    elif (rhead == _byte_swap_32(header_size)):
-        return 1, '>'
-    elif (rhead == header_size):
-        return 1, '<'
-    else:
-        raise RuntimeError("Incorrect Gadget format %s!" % str(rhead))
+    def validate(self):
+        """Validate data integrity."""
+        try:
+            self.open().close()
+            self.gadget_format
+            self.float_type
+        except:
```

Comment:
This will likely fail the lint checker as there is a rule about bare exceptions. You can get around this by catching `BaseException`.

Reply:
Since I see this usage at other places in the codebase, I guess this rule is somehow turned off. Do we want to enforce this rule and correct all other occurrences?

Reply:
I think I find the origin: http://flake8.pycqa.org/en/latest/user/error-codes.html

F707: an `except:` block as not the last exception handler

So for our use case it should be fine.

Reply:
You're right, we must be ignoring that PEP8 rule, although this was probably just to avoid having to fix it everywhere.

Reply:
Just a heads up, you don’t want to catch `BaseException`, that’s the same as bare except. Instead you want to do `except Exception`, see https://stackoverflow.com/questions/27995057/why-is-it-recommended-to-derive-from-exception-instead-of-baseexception-class-in

Reply:
I agree with Britton we should stop using bare except. That doesn’t need to be fixed in this PR.

Reply:
Okay, I'll leave this bare except here for future PR to fix then.